### ISSUE TITLE
[Feature] Add Fun-CosyVoice3 to SGLang-Omni.

### ISSUE THREAD

**Issue by PoTaTo-Mika** [sgl-project/sglang-omni#1302]
### Checklist

- [x] 1. If the issue you raised is not a feature but a question, please raise a discussion at https://github.com/sgl-project/sglang-omni/discussions/new/choose Otherwise, it will be closed.
- [x] 2. Please use English, otherwise it will be closed.

### Motivation

Fun-CosyVoice3 is a strong text-to-speech model with fine-grained control, multilingual synthesis, and few-shot voice cloning. It has been adopted by academic and industrial applications.

This issue proposes native support for `FunAudioLLM/Fun-CosyVoice3-0.5B-2512` in SGLang-Omni, including continuous batching for speech-token generation, streaming Flow + HiFT decoding, and the OpenAI-compatible `/v1/audio/speech` endpoint.

Tracing #1331 

### Related resources

- Hugging Face repository: https://huggingface.co/FunAudioLLM/Fun-CosyVoice3-0.5B-2512
- Official GitHub repository: https://github.com/QwenAudio/CosyVoice
- [SGLang-Omni TTS model integration guide](../developer_reference/tts_model_integration.md)
- [SGLang-Omni reference encode service](../developer_reference/reference_encode_service.md)

### Proposed pipeline

```text
preprocessing -> reference_encode -> tts_engine -> audio_decode
```

Implementation is split into two parallel workstreams:

```text
LLM:    text + LLM prompt tokens -> generated speech-token stream
Decode: reference audio + speech-token stream -> 24 kHz PCM stream
```

The initial API scope is:

| Inputs | Mode |
|---|---|
| `ref_audio` and `ref_text` | Zero-shot voice cloning |
| `ref_audio` without `ref_text` | Cross-lingual voice cloning |
| `ref_audio` and instructions | `instruct2` fine-grained control |

The checkpoint does not include `spk2info.pt`, so preset named speakers are not part of the initial scope. Voice conversion and the alternative `llm.rl.pt` checkpoint can be handled separately after the main TTS path is stable.

### Shared contract

The two workstreams should agree on this contract before implementation:

```python
@dataclass
class CosyVoice3ReferenceArtifact:
    llm_prompt_tokens: Tensor
    flow_prompt_tokens: Tensor
    prompt_mel: Tensor
    speaker_embedding: Tensor
    reference_text: str | None


@dataclass
class CosyVoice3TokenChunk:
    speech_tokens: Tensor  # Newly generated tokens only.
    is_final: bool
```

`llm_prompt_tokens` and `flow_prompt_tokens` must remain separate. Prompt speech tokens and prompt mel do not always have an exact 1:2 length ratio. The LLM should retain the complete prompt token sequence, while Flow must receive tokens and mel aligned to `prompt_mel_frames == 2 * flow_prompt_tokens`.

The decoder accumulates token deltas because causal Flow consumes the complete generated prefix on every invocation. Reference cache identity must include the model and encoder revision, preprocessing options, reference audio content identity, and reference text.

### LLM workstream

- [x] Add the Fun-CosyVoice3 Hugging Face config shim, pipeline config, and SGLang model registration.
- [x] Implement the native SGLang Qwen2 backbone: 24 layers, hidden size 896, 14 attention heads, 2 KV heads, head dimension 64, and intermediate size 4,864.
- [ ] Map `llm.pt` Q/K/V weights into `qkv_proj` and gate/up weights into `gate_up_proj`.
- [ ] Load the text embedding, Qwen2 backbone, `speech_embedding`, and 6,761-way `llm_decoder`.
- [ ] Skip the unused checkpoint text `lm_head`.
- [ ] Build prefill embeddings in the order `speech SOS + prompt/target text + speech task ID + LLM prompt speech tokens`.
- [ ] Implement RAS sampling with `top_p=0.8`, `top_k=25`, window size 10, and `tau_r=0.1`.
- [ ] Handle `SOS=6561`, `EOS=6562`, `task=6563`, `fill=6564`, and the stop range `6561..6760`.
- [ ] Stream newly generated speech tokens with a final marker.
- [ ] Keep sampling, seed, stop, and cancellation state request-scoped.
- [ ] Add CPU tests for config parsing, prompt construction, weight mapping, sampling, stop handling, and payload serialization.
- [ ] Add GPU parity tests for prefill logits, decode logits, and fixed-seed generated speech tokens.

### Decode workstream

- [ ] Implement reference audio loading, mono conversion, and 16/24 kHz resampling.
- [ ] Run `speech_tokenizer_v3.onnx` to extract 25 Hz prompt speech tokens.
- [ ] Run `campplus.onnx` to extract a 192-dimensional speaker embedding.
- [ ] Extract the checkpoint mel representation: 80 bins, `n_fft=1920`, `hop_size=480`, `win_size=1920`, `fmin=0`, `fmax=None`, and `center=False`.
- [ ] Align Flow prompt tokens and mel to the required 1:2 ratio without truncating the LLM prompt token sequence.
- [ ] Integrate the shared `ReferenceEncodeService` with content-safe cache keys and abort cleanup.
- [ ] Load all required weights from `flow.pt` and `hift.pt`.
- [ ] Implement the 22-layer, hidden-size-1,024 causal DiT with classifier-free guidance and 10 Euler steps.
- [ ] Preserve the 3-token Flow lookahead and 2 mel frames per speech token.
- [ ] Preserve HiFT's 24 kHz output and float64 F0 predictor until a parity test proves that lower precision is safe.
- [ ] Implement non-streaming token-to-wave decode first.
- [ ] Implement streaming state for accumulated tokens, `token_offset`, `current_hop`, accumulated mel, `speech_offset`, and final flush.
- [ ] Keep `current_hop` request-scoped; do not mutate a shared model attribute.
- [ ] Start with Flow request batch size 1 because upstream Flow internally expands the request to batch size 2 for classifier-free guidance.
- [ ] Add CPU tests for reference alignment, cache identity, state lifecycle, output length, final flush, and abort cleanup.
- [ ] Add GPU parity tests for Flow mel and HiFT audio behavior.

### Integration and validation

- [ ] Wire `preprocessing -> reference_encode -> tts_engine -> audio_decode`.
- [ ] Set `stream_to=["audio_decode"]` on `tts_engine` and allow `audio_decode` to accept stream chunks before the final payload.
- [ ] Add request normalization for zero-shot, cross-lingual, and `instruct2` modes.
- [ ] Add an example serving configuration and `/v1/audio/speech` usage documentation.
- [ ] Verify that completion, failure, cancellation, and abort all release LLM and decoder request state.
- [ ] Verify concurrent requests do not share hop, mel, waveform offset, sampler, or prompt state.
- [ ] Run multilingual and voice-cloning evaluation.
- [ ] Report WER/CER, speaker similarity, time to first audio, real-time factor, throughput, and GPU memory.

### Validated contracts

The following contracts were validated on CPU with the official checkpoint and the official 3.48-second `zero_shot_prompt.wav` clip:

| Component | Result |
|---|---|
| Speech tokenizer | 87 speech tokens |
| Prompt mel | `[1, 174, 80]` |
| CampPlus | `[1, 192]` |
| Flow final | 7 tokens -> 14 mel frames |
| Flow non-final | 7 tokens with 3-token lookahead -> 8 mel frames |
| HiFT final | 12 mel frames -> 5,760 PCM samples |
| HiFT non-final | Retains 8 mel frames / 3,840 PCM samples |
| Full decoder | 87 tokens -> 174 mel frames -> 83,520 PCM samples |

All 293 source tensors in `llm.pt` were audited. A total of 292 source tensors map to 172 logical SGLang parameters after QKV and gate/up packing. The only skipped tensor is the unused text `lm_head`. The mapped BF16 LLM weights occupy approximately 0.943 GiB.

Trimming one sample from the reference waveform changed the observed prompt lengths from `87 tokens / 174 mel frames` to `87 / 173`. This validates the separate LLM and Flow prompt-token fields in the shared contract.

For an 87-token streaming decoder run without a prompt prefix, the decoder emitted three PCM chunks of 20,160, 48,000, and 15,360 samples. Their total was 83,520 samples, equal to the non-streaming output length. Streaming tests should validate length, finite values, chunk-boundary continuity, speaker similarity, and ASR quality instead of waveform `allclose`.

### Acceptance criteria

- [ ] The checkpoint loads without missing required LLM, Flow, or HiFT weights.
- [ ] Zero-shot, cross-lingual, and `instruct2` requests work through `/v1/audio/speech`.
- [ ] Non-streaming responses contain valid mono 24 kHz audio with the expected duration.
- [ ] Streaming emits audio before speech-token generation completes and preserves the final sample count.
- [ ] Reference cache keys cannot alias different audio with the same transcript.
- [ ] Concurrent and aborted requests leave no shared or stale request state.
- [ ] Fixed-seed generation and decode satisfy the agreed upstream parity tolerances.
- [ ] Quality and performance results are included in the implementation PR.

### Open questions

- [ ] Confirm full LLM prompt tokens plus aligned Flow prompt tokens against zero-shot quality baselines. This matches the newer upstream Triton runtime; the older frontend trims both sequences.
- [ ] Decide between minimally vendoring the required model classes and using a narrow CosyVoice compatibility layer. The full upstream dependency set conflicts with SGLang-Omni's Torch and Transformers versions.
- [ ] Profile Flow and HiFT in one `audio_decode` process before considering a split. Their streaming state should remain colocated initially.
- [ ] Test whether the float64 F0 predictor can be optimized without audible regressions.
- [ ] Determine GPU placement and memory fractions for the approximately 0.5B LLM, 332M Flow, and 21M HiFT components.

**Comment by PoTaTo-Mika** [comment]
closed as finished.
